subnet/subnet_contract.go

```diff
@@ -111,27 +163,437 @@ func Delete(id []byte) {
+		panic("addUser: " + ErrAccessDenied)
+	}
+
+	stKey[0] = userPrefix
+
+	if keyInList(ctx, user, stKey) {
+		panic("addUser: user has already been added")
+	}
+
+	putKeyInList(ctx, user, stKey)
+}
+
+// RemoveUser removes user from the specified subnetwork and group.
+// Must be called by the owner or the group's admin only.
+func RemoveUser(subnetID []byte, groupID []byte, user []byte) {
+	// V2 format check
+	if len(subnetID) != subnetIDSize {
+		panic("removeUser: " + ErrInvalidSubnetID)
+	}
+
+	// V2 format check
+	if len(groupID) != groupIDSize {
+		panic("removeUser: " + ErrInvalidGroupID)
+	}
+
+	ctx := storage.GetContext()
+
+	stKey := append([]byte{ownerPrefix}, subnetID...)
+
+	rawOwner := storage.Get(ctx, stKey)
+	if rawOwner == nil {
+		panic("removeUser: " + ErrNotExist)
+	}
+
+	stKey[0] = clientAdminPrefix
+	stKey = append(stKey, groupID...)
+
+	owner := rawOwner.([]byte)
+
+	if !calledByOwnerOrAdmin(ctx, owner, stKey) {
+		panic("removeUser: " + ErrAccessDenied)
+	}
+
+	stKey[0] = userPrefix
+
+	if !keyInList(ctx, user, stKey) {
+		panic("removeUser: " + ErrUserNotExist)
+	}
+
+	deleteKeyFromList(ctx, user, stKey)
+}
+
+// UserAllowed returns bool that indicates if node is included in the
+// specified subnet or not.
+func UserAllowed(subnetID []byte, user []byte) bool {
+	// V2 format check
+	if len(subnetID) != subnetIDSize {
+		panic("userAllowed: " + ErrInvalidSubnetID)
+	}
+
+	ctx := storage.GetContext()
+
+	stKey := append([]byte{ownerPrefix}, subnetID...)
+	if storage.Get(ctx, stKey) == nil {
+		panic("userAllowed: " + ErrNotExist)
+	}
+
+	stKey[0] = userPrefix
+	prefixLen := len(stKey) + groupIDSize
+
+	iter := storage.Find(ctx, stKey, storage.KeysOnly)
+	for iterator.Next(iter) {
+		key := iterator.Value(iter).([]byte)
+		if common.BytesEqual(user, key[prefixLen:]) {
+			return true
+		}
+	}
+
+	return false
 }
 
 // Version returns version of the contract.
 func Version() int {
 	return common.Version
 }
+
+func getPublicKeysByPrefix(ctx storage.Context, prefix []byte) []interop.PublicKey {
+	var keys []interop.PublicKey
+	prefixLen := len(prefix)
+
+	iter := storage.Find(ctx, prefix, storage.KeysOnly)
+	for iterator.Next(iter) {
+		key := iterator.Value(iter).([]byte)
+		keys = append(keys, key[prefixLen:])
+	}
+
+	return keys
+}
+
+func keyInList(ctx storage.Context, searchedKey interop.PublicKey, prefix []byte) bool {
+	return storage.Get(ctx, append(prefix, searchedKey...)) != nil
+}
+
+func putKeyInList(ctx storage.Context, keyToPut interop.PublicKey, prefix []byte) {
+	storage.Put(ctx, append(prefix, keyToPut...), []byte{1})
+}
+
+func deleteKeyFromList(ctx storage.Context, keyToDelete interop.PublicKey, prefix []byte) {
+	storage.Delete(ctx, append(prefix, keyToDelete...))
+}
+
+func calledByOwnerOrAdmin(ctx storage.Context, owner []byte, adminPrefix []byte) bool {
+	if runtime.CheckWitness(owner) {
+		return true
+	}
+
+	prefixLen := len(adminPrefix)
+
+	iter := storage.Find(ctx, adminPrefix, storage.KeysOnly)
```

Comment:
```suggestion
	iter := storage.Find(ctx, adminPrefix, storage.KeysOnly | storage.RemovePrefix)
```

This way you won't need to cut prefix manually.

Reply:
> // RemovePrefix is used for stripping 1-byte prefix from keys.
	RemovePrefix FindFlags = 1 << 1

are you sure that it is a suitable flag for that?

Reply:
Oups, missed that.